registry/storage/driver/s3-aws/s3.go

```diff
@@ -561,45 +561,57 @@ func (d *driver) Move(ctx context.Context, sourcePath string, destPath string) e
 	return d.Delete(ctx, sourcePath)
 }
 
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
 // Delete recursively deletes all objects stored at "path" and its subpaths.
+// We must be careful since S3 does not guarantee read after delete consistency
 func (d *driver) Delete(ctx context.Context, path string) error {
-	resp, err := d.S3.ListObjects(&s3.ListObjectsInput{
+	s3Objects := make([]*s3.ObjectIdentifier, 0, listMax)
+	listObjectsInput := &s3.ListObjectsInput{
 		Bucket: aws.String(d.Bucket),
 		Prefix: aws.String(d.s3Path(path)),
-	})
-	if err != nil || len(resp.Contents) == 0 {
-		return storagedriver.PathNotFoundError{Path: path}
 	}
+	for {
+		// list all the objects
+		resp, err := d.S3.ListObjects(listObjectsInput)
 
-	s3Objects := make([]*s3.ObjectIdentifier, 0, listMax)
+		// resp.Contents can only be empty on the first call, because we wouldnt recall ListObjects if there are no more results
+		if err != nil || len(resp.Contents) == 0 {
```

Comment:
I spent quite a bit of time thinking about our loop conditions, and came to this conclusion. I would appreciate if you could double check my logic below to make sure it is correct.

From the docs, IsTruncated "specifies whether (true) or not (false) all of the results were returned. If the number of results exceeds that specified by MaxKeys, all of the results might not be returned." From what I understand, this means that IsTruncated will only be true if there are more objects that have not been returned. Therefore, I believe we would only loop again if there are more objects. If no objects are returned, either it is the first iteration (Not Found) or something weird is happening, which is an error.

Here are the docs for reference: http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html